docs/internal/vmmap.md

```diff
@@ -0,0 +1,272 @@
+# Memory Management and Vmmap
+
+## What is a Vmmap?
+
+A vmmap is a tool for managing the memory layout of a process within an operating system. It provides detailed information about the various memory regions allocated to the process, including the heap, stack, and memory-mapped files. The vmmap also displays the access permissions for each region, specifying whether the memory is read-only, read-write, or executable. Additionally, it highlights the start and end addresses of each region, their sizes, and any files mapped into the process's address space.
+
+## Motivation
+
+Wasmtime traditionally manages memory using WebAssembly's linear memory model, where each instance is allocated a contiguous block of memory divided into pages (typically 64 KiB). This memory can grow or shrink dynamically in page-sized increments, subject to limits defined by the WebAssembly module or system constraints. Because Lind emulates processes as cages within a single address space, we need to be able to track valid memory regions within each cage. We've integrated a vmmap system into Lind that manages memory more like POSIX-based systems. This addition enables proper implementation of syscalls such as brk(), mmap(), and mprotect() for allocation, deallocation, and permission management of memory.
+
+### Necessary For:
+
+#### fork()
+The fork() system call necessitates copying the memory space of the parent process to the child. However, accurately replicating the parent's memory in the child requires tracking the memory protections of each region and identifying which memory regions are mapped as shared. Without this information, a precise representation of the parent's memory in the child cannot be ensured.
+
+#### malloc()
+Glibc's implementation of malloc relies on a POSIX style heap, which is contiguous. Using a "greedy" memory allocation style would in many cases break up the heap causing malloc() to fail.
```

Comment:
malloc isn't a system call.   How would an OS level concept like the VMmap interact?

Reply:
I'm confused by how this question is phrased? Do you mean how does the vmmap interact with system calls?

Reply:
I'm just a little confused about why we're talking about malloc if we're operating at the OS layer.   It's not a concept that is visible to us.   We see brk, mmap, etc.